### PR TITLE
Shared definition for common transducer init and result functions

### PR DESCRIPTION
**Pull request by paldepind**
As per the discussion in #1020.

I don't think I've missed anything :)

### CODE REVIEW

**Comment by CrossEye** [comment]
While I'm probably :+1: on this change, I've got to say that seeing all those `prototypes` and constructor functions in our FP library is not a good feeling!

But if we're going to do it, are we better off going just a little further and adding something like `inheritFrom`?:

```javascript
XAll.prototype = _inheritFrom(_xfBase);
XAll.prototype['@@transducer/result'] = function(result) {...}
```

**Comment by buzzdecafe** [comment]
inheritance seems like overkill to me, since all we're doing is sharing function references.

**Comment by davidchambers** [comment]
> inheritance seems like overkill to me, since all we're doing is sharing function references.

:+1:

**Comment by buzzdecafe** [comment]
we all good then? I say :cow2:

**Comment by CrossEye** [comment]
> inheritance seems like overkill to me, since all we're doing is sharing function references.

What else would you mean by inheritance?   Especially in JS.

But yes, I'm good to go.

File: src/internal/_xfBase.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+module.exports = {
+   init: function() {
```
Comment (davidchambers) [inline src/internal/_xfBase.js]: :triangular_ruler: You have three spaces here rather than two.

**Comment by davidchambers** [comment]
> What else would you mean by inheritance?

An approach which enabled `dog instanceof Animal` to evaluate true.

**Comment by CrossEye** [comment]
I almost never use `instanceof` in JS, preferring to use duck-typing.  But yes, you do want to know something about object types.   I would argue, though that you only need that information as a proxy for the types of messages you can send to that object: in JS those are the function properties it has along its prototype chain.

**Comment by kevinbeaty** [comment]
What about something like this (not sure on the naming):

```javascript
function _xfDefaults(xfImpl){
   xfImpl.prototype['@@transducer/init'] = _xfInit
   xfImpl.prototype['@@transducer/result'] = _xfStep
   // ? return xfImpl
}

// later
function XTake(n){}
_xfDefaults(XTake)
XTake.prototype['@@transducer/step'] = ...

// or even
function _xfWithStep(xfImpl, xfStep){
   xfImpl.prototype['@@transducer/init'] = _xfInit
   xfImpl.prototype['@@transducer/step'] = _xfStep
   xfImpl.prototype['@@transducer/result'] = _xfStep
   // ? return xfImpl
}


_xfWithStep(XTake, function(acc, value){})
```

Just a  thought.  May not be worth it.  I'm not opposed to the current PR either.

**Comment by CrossEye** [comment]
@kevinbeaty: Your first suggestion is similar to what I was thinking about.  I'm perfectly happy for the moment with this PR.  Whatever unhappiness I have is with the usage of so many constructor functions and prototypes to implement this part of our functional library.  Something really seems off.

**Comment by paldepind** [comment]
@kevinbeaty: That's not a bad idea. But it may not be worth since were only adding 1 or 2 functions (as far as I know there are no step function shared across the transducers). If this is what people prefer I can change this PR.

@carlosrocha: Getting rid of the the constructor functions and prototype usage is certainly do able. `_xfind` for instance could be changed from this:

```javascript
module.exports = (function() {
  function XFind(f, xf) {
    this.xf = xf;
    this.f = f;
    this.found = false;
  }
  XFind.prototype['@@transducer/init'] = _xfBase.init;
  XFind.prototype['@@transducer/result'] = function(result) {
    if (!this.found) {
      result = this.xf['@@transducer/step'](result, void 0);
    }
    return this.xf['@@transducer/result'](result);
  };
  XFind.prototype['@@transducer/step'] = function(result, input) {
    if (this.f(input)) {
      this.found = true;
      result = _reduced(this.xf['@@transducer/step'](result, input));
    }
    return result;
  };

  return _curry2(function _xfind(f, xf) { return new XFind(f, xf); });
})();
```

To something like this:

```javascript
var resFn = function(result) {
  if (!this.found) {
    result = this.xf['@@transducer/step'](result, void 0);
  }
  return this.xf['@@transducer/result'](result);
};
var stepFn = function(result, input) {
  if (this.f(input)) {
    this.found = true;
    result = _reduced(this.xf['@@transducer/step'](result, input));
  }
  return result;
};
module.exports = curry2(function(f, xf) {
  return {
    xf: xf,
    f: f,
    found: false,
    '@@transducer/init': _xfBase.init,
    '@@transducer/result': resFn,
    '@@transducer/step': stepFn,
  }
});
```

Do you like the second version better?

**Comment by buzzdecafe** [comment]
unless there is some demonstrable benefit from this rewrite, i don't think it's worth it. the interface is the point, not the implementation.

**Comment by CrossEye** [comment]
> Do you like the second version better?

Assuming that Carlos was some weird typo for me (:smile:), then yes I do, but not enough to make it worth going back and changing the existing implementations unless we're in there doing other things.  I agree with @buzzdecafe here.

**Comment by paldepind** [comment]
> Assuming that Carlos was some weird typo for me (:smile:)

Yacks! I apparently can't handle the GitHub autocompletion :(

I'm indifferent about this – and happy as long as the duplication goes away.

**Comment by CrossEye** [comment]
So is everyone on-board with this?

**Pull request merged**